File: windows/src/core/qm-decoder.js
"use strict";

const fs = require("node:fs");
const path = require("node:path");
const { encryptedExtensions, privateKey256, extensionOf } = require("./constants");
const { QMCKeyDecoder } = require("./qmc-key-decoder");
const { QMStaticCipher, QMMapCipher, QMRC4Cipher } = require("./qmc-cipher");

function isValidAudioHeader(data, fileExtension) {
  if (!data || data.length === 0) return false;
  const head = data.subarray(0, 16);
  switch (fileExtension) {
    case "flac":
      return head.subarray(0, 4).toString("ascii") === "fLaC";
    case "ogg":
      return head.subarray(0, 4).toString("ascii") === "OggS";
    case "mp3":
      return head.subarray(0, 3).toString("ascii") === "ID3" || (head.length >= 2 && head[0] === 0xff && (head[1] & 0xe0) === 0xe0);
    case "m4a":
      return head.length >= 12 && head.subarray(4, 8).toString("ascii") === "ftyp";
    case "wav":
      return head.subarray(0, 4).toString("ascii") === "RIFF";
    default:
      return true;
  }
}

class QMDecoder {
  constructor(originFilePath) {
    this.originFilePath = originFilePath;
    this.fileBuffer = fs.readFileSync(originFilePath);
    this.originFileLength = this.fileBuffer.length;
    this.encryptedAudioLength = this.originFileLength;
    this.realAudioSize = 0;
    this.cipher = null;
    this.searchKey();
  }

  decryptAudioData() {
    const ext = extensionOf(this.originFilePath);
    const mapped = encryptedExtensions[ext];
    if (!mapped || !this.cipher) throw new Error(`不支持的 QMC 扩展名：.${ext}`);
    const audioData = this.fileBuffer.subarray(0, this.realAudioSize);
    const decoded = this.cipher.decrypt(audioData, 0);
    if (!isValidAudioHeader(decoded, mapped.ext)) {
      throw new Error("无法解密：当前文件可能需要 ekey，或此加密版本暂不支持。");
    }
    return { data: decoded, fileExtension: mapped.ext };
  }

  searchKey() {
    this.encryptedAudioLength = this.detectEncryptedAudioLength();
    if (this.encryptedAudioLength < 4) throw new Error("文件太小，无法读取 key。");
    const lastFourBytes = this.fileBuffer.subarray(this.encryptedAudioLength - 4, this.encryptedAudioLength);

    if (lastFourBytes.toString("utf8") === "QTag") {
      if (this.originFileLength < 8) throw new Error("无法读取 QTag key 长度。");
      const keySize = this.fileBuffer.readUInt32BE(this.originFileLength - 8);
      this.realAudioSize = this.originFileLength - keySize - 8;
      if (this.realAudioSize < 0) throw new Error("QTag key 长度异常。");
      const rawKey = this.fileBuffer.subarray(this.realAudioSize, this.realAudioSize + keySize);
      const keyEndIndex = rawKey.indexOf(0x2c);
      if (keyEndIndex < 0) throw new Error("无法定位 QTag key 结束位置。");
      this.setCipher(rawKey.subarray(0, keyEndIndex));
      return;
    }

    const keySize = lastFourBytes.readUInt32LE(0);
    if (keySize < 0x300) {
      this.realAudioSize = this.encryptedAudioLength - keySize - 4;
      if (this.realAudioSize < 0) throw new Error("key 长度异常。");
      const rawKey = this.fileBuffer.subarray(this.realAudioSize, this.realAudioSize + keySize);
      this.setCipher(rawKey);
    } else {
      this.realAudioSize = this.encryptedAudioLength;
      this.cipher = new QMStaticCipher(privateKey256);
    }
  }

  setCipher(keyBuffer) {
    const keyDecoder = new QMCKeyDecoder();
    const decodedKey = keyDecoder.deriveKey(keyBuffer);
    this.cipher = decodedKey.length > 300 ? new QMRC4Cipher(decodedKey) : new QMMapCipher(decodedKey);
  }

  detectEncryptedAudioLength() {
    if (this.originFileLength < 16) return this.originFileLength;
    const tailMagic = this.fileBuffer.subarray(this.originFileLength - 8).toString("utf8");
    if (tailMagic !== "musicex\0") return this.originFileLength;
    const trailerSize = this.fileBuffer.readUInt32LE(this.originFileLength - 16);
    if (trailerSize <= 0 || trailerSize >= this.originFileLength) return this.originFileLength;
    return this.originFileLength - trailerSize;
  }
}

function baseNameWithoutExtension(filePath) {
  return path.basename(filePath, path.extname(filePath));
}

module.exports = { QMDecoder, baseNameWithoutExtension, isValidAudioHeader };

File: autoMC/QMDecoder.swift
//
//  QMCDecoder.swift
//  QMCDecode
//
//  Created by 龚杰洪 on 2022/1/10.
//  Copyright © 2022 龚杰洪. All rights reserved.
//

import Foundation

class QMDecoder {
    enum DecoderError: Error {
        case unsupportFileExtension(ext: String)
        case canNotReadFile
        case canNotReadFileByStream
        case canNotGetFileLength
        case canNotReadSizeBuffer
        case canNotReadRawKeyBuffer
        case searchRawKeyFailed
        case unsupportedEncryptedKey
    }

    private let commaASCIICode: UInt8 = Character(",").asciiValue ?? 44


    private let originFilePath: String
    private let outputDirectory: String
    private let readStream: InputStream
    private let originFileLength: Int
    private var encryptedAudioLength: Int

    private var realAudioSize: Int = 0

    private var cipher: QMCipher?

    init(originFilePath: String, outputDirectory: String) throws {
        self.originFilePath = originFilePath
        self.outputDirectory = outputDirectory
        guard let fileStream = InputStream(fileAtPath: originFilePath) else {
            throw DecoderError.canNotReadFileByStream
        }
        self.readStream = fileStream

        let fileAttributes = try FileManager.default.attributesOfItem(atPath: originFilePath)
        guard let fileLength = fileAttributes[FileAttributeKey.size] as? Int else {
            throw DecoderError.canNotGetFileLength
        }
        self.originFileLength = fileLength
        self.encryptedAudioLength = fileLength

        try searchKey()
    }

    @discardableResult
    func decryptAndWriteToFile(outputExtension: String? = nil) throws -> URL {
        let decrypted = try decryptedAudioData()
        let fileURL = URL(fileURLWithPath: originFilePath)
        let targetExtension = outputExtension ?? decrypted.fileExtension

        var outputURL = URL(fileURLWithPath: self.outputDirectory)
        outputURL.appendPathComponent(fileURL.lastPathComponent)
        outputURL.deletePathExtension()
        outputURL.appendPathExtension(targetExtension)
        outputURL = uniqueOutputURL(for: outputURL)
        try decrypted.data.write(to: outputURL, options: Data.WritingOptions.atomic)
        return outputURL
    }

    func decryptedAudioData() throws -> (data: Data, fileExtension: String) {
        let fileURL = URL(fileURLWithPath: originFilePath)
        let fileExtension = fileURL.pathExtension.lowercased()
        if fileExtension.count > 0, let extAndVersion = encryptExtDictionary[fileExtension], let cipher = self.cipher {
            let fileHandle = FileHandle(forReadingAtPath: originFilePath)
            if let fileData = try fileHandle?.read(upToCount: self.realAudioSize) {
                let decodeData = cipher.qmDecrypt(data: fileData, offset: 0)
                guard isValidAudioHeader(decodeData, fileExtension: extAndVersion.ext) else {
                    throw DecoderError.unsupportedEncryptedKey
                }
                return (decodeData, extAndVersion.ext)
            } else {
                throw DecoderError.canNotReadFile
            }
        } else {
            throw DecoderError.unsupportFileExtension(ext: fileExtension)
        }
    }

    func matchingDecoder(_ extAndVersion: ExtensionAndVersion) throws {
        if extAndVersion.version == .v2 {

        } else {

        }
    }

    func searchKey() throws {
        guard let fileHandle = FileHandle(forReadingAtPath: originFilePath) else {
            throw DecoderError.canNotReadFile
        }
        defer {
            try? fileHandle.close()
        }

        self.encryptedAudioLength = try detectEncryptedAudioLength(fileHandle: fileHandle)

        try fileHandle.seek(toOffset: UInt64(self.encryptedAudioLength - 4))
        guard let lastFourBytes = try fileHandle.read(upToCount: 4) else {
            throw DecoderError.canNotReadFile
        }

        // 移动端下载的用,以QTag结尾
        if String(bytes: lastFourBytes, encoding: String.Encoding.utf8) == "QTag" {
            // 读取key长度
            try fileHandle.seek(toOffset: UInt64(self.originFileLength - 8))
            guard let sizeBuffer = try fileHandle.read(upToCount: 4) else {
                throw DecoderError.canNotReadFile
            }
            let keySize = sizeBuffer.withUnsafeBytes {
                $0.load(as: UInt32.self).bigEndian
            }

            // 计算真实音频长度
            self.realAudioSize = self.originFileLength - Int(keySize) - 8

            // 读取原始key
            try fileHandle.seek(toOffset: UInt64(self.realAudioSize))
            guard let rawKey = try fileHandle.read(upToCount: Int(keySize)) else {
                throw DecoderError.canNotReadRawKeyBuffer
            }

            // 通过逗号找到key结束位置
            guard let keyEndIndex = rawKey.firstIndex(of: commaASCIICode) else {
                throw DecoderError.searchRawKeyFailed
            }

            // 通过原始key和key结束位置组装解码器
            try setCipher(keyBuffer: [UInt8]([UInt8](rawKey)[0..<keyEndIndex]))
        } else {
            // PC macOS端下载的文件
            let keySize = lastFourBytes.withUnsafeBytes {
                $0.load(as: UInt32.self).littleEndian
            }

            if keySize < 0x300 {
                // key 在固定位置
                self.realAudioSize = self.encryptedAudioLength - Int(keySize) - 4
                try fileHandle.seek(toOffset: UInt64(self.realAudioSize))
                guard let rawKey = try fileHandle.read(upToCount: Int(keySize)) else {
                    throw DecoderError.canNotReadRawKeyBuffer
                }

                try setCipher(keyBuffer: [UInt8](rawKey))
            } else {
                // 用固定key解码
                self.realAudioSize = self.encryptedAudioLength
                self.cipher = try QMStaticCipher(originKey: privateKey256)
            }
        }
    }

    func setCipher(keyBuffer: [UInt8]) throws {
        let keyDecoder = QMCKeyDecoder()
        let decodedKey = try keyDecoder.deriveKey(keyBuffer)

        if decodedKey.count > 300 {
            self.cipher = try QMRC4Cipher(originKey: decodedKey)
        } else {
            self.cipher = try QMMapCipher(originKey: decodedKey)
        }
    }

    private func uniqueOutputURL(for url: URL) -> URL {
        let fileManager = FileManager.default
        guard fileManager.fileExists(atPath: url.path) else {
            return url
        }

        let directory = url.deletingLastPathComponent()
        let baseName = url.deletingPathExtension().lastPathComponent
        let fileExtension = url.pathExtension

        var index = 2
        while true {
            let candidate = directory
                .appendingPathComponent("\(baseName) \(index)")
                .appendingPathExtension(fileExtension)
            if fileManager.fileExists(atPath: candidate.path) == false {
                return candidate
            }
            index += 1
        }
    }

    private func detectEncryptedAudioLength(fileHandle: FileHandle) throws -> Int {
        guard originFileLength >= 16 else {
            return originFileLength
        }

        try fileHandle.seek(toOffset: UInt64(originFileLength - 8))
        guard let tailMagic = try fileHandle.read(upToCount: 8),
              String(data: tailMagic, encoding: .utf8) == "musicex\0" else {
            return originFileLength
        }

        try fileHandle.seek(toOffset: UInt64(originFileLength - 16))
        guard let sizeData = try fileHandle.read(upToCount: 4), sizeData.count == 4 else {
            return originFileLength
        }

        let trailerSize = sizeData.withUnsafeBytes {
            Int($0.load(as: UInt32.self).littleEndian)
        }
        guard trailerSize > 0, trailerSize < originFileLength else {
            return originFileLength
        }

        return originFileLength - trailerSize
    }

    private func isValidAudioHeader(_ data: Data, fileExtension: String) -> Bool {
        let bytes = [UInt8](data.prefix(16))
        guard bytes.isEmpty == false else {
            return false
        }

        switch fileExtension {
        case "flac":
            return bytes.starts(with: Array("fLaC".utf8))
        case "ogg":
            return bytes.starts(with: Array("OggS".utf8))
        case "mp3":
            return bytes.starts(with: Array("ID3".utf8))
                || (bytes.count >= 2 && bytes[0] == 0xff && (bytes[1] & 0xe0) == 0xe0)
        case "m4a":
            return bytes.count >= 12 && bytes[4...7].elementsEqual(Array("ftyp".utf8))
        case "wav":
            return bytes.starts(with: Array("RIFF".utf8))
        default:
            return true
        }
    }
}

extension QMDecoder.DecoderError: LocalizedError {
    var errorDescription: String? {
        switch self {
        case .unsupportFileExtension(let ext):
            return L10n.tr("error.unsupportedExtension", ext)
        case .canNotReadFile:
            return L10n.tr("error.cannotReadFile")
        case .canNotReadFileByStream:
            return L10n.tr("error.cannotOpenStream")
        case .canNotGetFileLength:
            return L10n.tr("error.cannotGetLength")
        case .canNotReadSizeBuffer:
            return L10n.tr("error.cannotReadSize")
        case .canNotReadRawKeyBuffer:
            return L10n.tr("error.cannotReadRawKey")
        case .searchRawKeyFailed:
            return L10n.tr("error.searchRawKeyFailed")
        case .unsupportedEncryptedKey:
            return L10n.tr("error.unsupportedEncryptedKey")
        }
    }
}
